### DISCUSSION TITLE
How to report a dependency issue with CefSharp.Core.Runtime?

### DISCUSSION THREAD

**Discussion by rboberholz** [cefsharp/CefSharp#4468] [General]
DependancyWalker.Net application reports:
CefSharp.Core--->CefSharp.Core.Runtime - Error loading
System.BadImageFormatException: Could not load file or assembly 'CefSharp.Core.Runtime, Version=112.2.70.0, Culture=neutral, PublicKeyToken=40c4b6fc221f4138' or one of its dependencies. An attempt was made to load a program with an incorrect format.
File name: 'CefSharp.Core.Runtime, Version=112.2.70.0, Culture=neutral, PublicKeyToken=40c4b6fc221f4138' ---> System.BadImageFormatException: Could not load file or assembly 'file:///F:\_OneDriveBusiness\Intent Digital L.L.C\GUS CEF - Source Code\VoteLynx_VP_CEF\bin\Debug\CefSharp.Core.Runtime.dll' or one of its dependencies. An attempt was made to load a program with an incorrect format.
File name: 'file:///F:\_OneDriveBusiness\Intent Digital L.L.C\GUS CEF - Source Code\VoteLynx_VP_CEF\bin\Debug\CefSharp.Core.Runtime.dll'
   at System.Reflection.RuntimeAssembly._nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, Evidence assemblySecurity, RuntimeAssembly reqAssembly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadFrom(String assemblyFile, Evidence securityEvidence, Byte[] hashValue, AssemblyHashAlgorithm hashAlgorithm, Boolean forIntrospection, Boolean suppressSecurityChecks, StackCrawlMark& stackMark)
   at System.Reflection.Assembly.LoadFrom(String assemblyFile)
   at SindaSoft.DependencyWalker.Walker.currentDomain_AssemblyResolve(Object sender, ResolveEventArgs e)
   at System.AppDomain.OnAssemblyResolveEvent(RuntimeAssembly assembly, String assemblyFullName)

This same dependency showed up on occasion in an app on loading CefSharp.WinForms.dll.  Checking this .dll with DependencyWalker.net, it reported that the CefSharp.Core.Runtime.dll has the problem. This was first encountered with version 111.2.70.
I filled out information per the template instructions according to https://github.com/cefsharp/CefSharp/blob/master/.github/ISSUE_TEMPLATE/bug_report.md#bug-report.
 I believe one submits a bug report to StackOverflow. However, nothing told me how to submit the report to StackOverflow.
You need better instructions.
Thanks

**Comment by amaitland** [comment]
Please use [Discussions](https://github.com/cefsharp/CefSharp/discussions) to ask support related questions.

> System.BadImageFormatException: Could not load file or assembly 'CefSharp.Core.Runtime, Version=112.2.70.0, Culture=neutral, PublicKeyToken=40c4b6fc221f4138' or one of its dependencies. An attempt was made to load a program with an incorrect format.

This is a common exception and unlikely a bug in `CefSharp`. 

`An attempt was made to load a program with an incorrect format.` usually means you are loading the incorrect architecture dll.

> Checking this .dll with DependencyWalker.net, it reported that the CefSharp.Core.Runtime.dll has the problem.

Reported what problem exactly? 

> This was first encountered with version 111.2.70.

Are you upgrading from a previous version? No new dependencies have been added in quite some time.

Are you seeing this problem on your development machine or when you deploy to machine?

Please answer the questions at https://github.com/cefsharp/CefSharp/wiki/Questions-&-Support#when-asking-a-question-please-include-the-following

Need more information to provide any sort of guidance here. The [MinimalExample](https://github.com/cefsharp/CefSharp.MinimalExample) is kept up to date for testing purposes.

**Comment by rboberholz** [comment]
Alex,

I did use ‘DependanceWalker.net’ (DependencyWalker.exe) on the CefSharp.Cor.Runtime.dll and this is what is reported as an error:

F:\_OneDriveBusiness\Intent Digital L.L.C\GUS CEF - Source Code - Copy\VoteLynx_VP_CEF\bin\Debug\CefSharp.Core.Runtime.dll - Error loading
System.BadImageFormatException: Could not load file or assembly 'file:///F:\_OneDriveBusiness\Intent Digital L.L.C\GUS CEF - Source Code - Copy\VoteLynx_VP_CEF\bin\Debug\CefSharp.Core.Runtime.dll' or one of its dependencies. An attempt was made to load a program with an incorrect format.
File name: 'file:///F:\_OneDriveBusiness\Intent Digital L.L.C\GUS CEF - Source Code - Copy\VoteLynx_VP_CEF\bin\Debug\CefSharp.Core.Runtime.dll'
   at System.Reflection.RuntimeAssembly._nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, Evidence assemblySecurity, RuntimeAssembly reqAssembly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadFrom(String assemblyFile, Evidence securityEvidence, Byte[] hashValue, AssemblyHashAlgorithm hashAlgorithm, Boolean forIntrospection, Boolean suppressSecurityChecks, StackCrawlMark& stackMark)
   at System.Reflection.Assembly.LoadFrom(String assemblyFile)
   at SindaSoft.DependencyWalker.Walker.inspectAssembly(String name)


SEE BELOW: My responses highlighted in yellow.

Thank You,

Richard B. Oberholz
Engineer
Intent Digital
602-404-0100 Ext: 25
Email: ***@***.******@***.***>
Web:  www.intentdigital.com<http://www.intentdigital.com/>
             www.votelynx.com<http://www.votelynx.com/>

**Reply by amaitland** [reply]
> I did use ‘DependanceWalker.net’ (DependencyWalker.exe) on the CefSharp.Cor.Runtime.dll and this is what is reported as an error:

I'm not exactly sure which DependanceWalker project you are referring to exactly. As `CefSharp.Core.Runtime.dll` is a `Mixed Mode CLI/C++` assembly it's not your typical `managed` assembly. If you are seeing the same exception as when running your application then I'd be inclined to ignore the DependanceWalker output.

> SEE BELOW: My responses highlighted in yellow.

`GitHub` doesn't support colours. Your responses are hard to read as a result. In future please login to `GitHub` to respond, then you can use markdown formatting.

> What dependencies does the CefSharp.Core.Runtime

The [Readme.txt](https://github.com/cefsharp/CefSharp/blob/cefsharp/112/NuGet/Readme.txt#L27) file that is presented when you installed the Nuget package has a basic list. There is a more detailed list at https://github.com/cefsharp/CefSharp/wiki/Output-files-description-table-%28Redistribution%29

> what is the correct format?

You've not included that information so I cannot say. The answer depends on if you are using `AnyCPU`, `x64`, `x86`.

If you haven't already I'd suggest you read https://learn.microsoft.com/en-us/dotnet/api/system.badimageformatexception?view=net-7.0 for background.

> Sometimes it would show on the development machine, and I don’t know why it’s intermittent. The app with CefSharp would run fine after an initial installation. However, if you copied an updated .exe, after a minor change, to the target machine, the error would show up.

I'd need more information e.g. an example that reproduces the problem (no zip files) to know what's going on.

**Reply by amaitland** [reply]
> Need more information to provide any sort of guidance here. The [MinimalExample](https://github.com/cefsharp/CefSharp.MinimalExample) is kept up to date for testing purposes.

Does the problem reproduce using the `MinimalExample`?

**Comment by rboberholz** [comment]
Alex,

Get the DependancyWalker.net app from isindicic/DependencyWalker.Net: Show dependencies tree for .NET assemblies like old Depend Walker show it for non managed applications. (github.com)<https://github.com/isindicic/DependencyWalker.Net> and run it against the CefSharp.Core.Runtime.dll. Its found on GitHub at https://github.com/isindicic/DependencyWalker.Net and the above link. I believe I found this utility through Stackoverflow.com when trying to get answers to fix the problem.

I could not duplicate the problem using the ‘MinimalExample’.

If you have more questions, feel free to call me at the number listed below.

Clarification to an earlier email from today.
GH question: An attempt was made to load a program with an incorrect format. usually means you are loading the incorrect architecture dll.

RBO answer: The app architecture using CefSharp is set for 64bit. It also uses .net framework version 4.7.2. Is there a different architecture dll you are referring to.
Checking this .dll with DependencyWalker.net, it reported that the CefSharp.Core.Runtime.dll has the problem.

-----------------------------------------------------------------
GH question: Reported what problem exactly?

RBO answer:
F:\_OneDriveBusiness\Intent Digital L.L.C\GUS CEF - Source Code - Copy\VoteLynx_VP_CEF\bin\Debug\CefSharp.Core.Runtime.dll - Error loading
System.BadImageFormatException: Could not load file or assembly 'file:///F:\_OneDriveBusiness\Intent Digital L.L.C\GUS CEF - Source Code - Copy\VoteLynx_VP_CEF\bin\Debug\CefSharp.Core.Runtime.dll' or one of its dependencies. An attempt was made to load a program with an incorrect format.
File name: 'file:///F:\_OneDriveBusiness\Intent Digital L.L.C\GUS CEF - Source Code - Copy\VoteLynx_VP_CEF\bin\Debug\CefSharp.Core.Runtime.dll'
   at System.Reflection.RuntimeAssembly._nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, Evidence assemblySecurity, RuntimeAssembly reqAssembly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadFrom(String assemblyFile, Evidence securityEvidence, Byte[] hashValue, AssemblyHashAlgorithm hashAlgorithm, Boolean forIntrospection, Boolean suppressSecurityChecks, StackCrawlMark& stackMark)
   at System.Reflection.Assembly.LoadFrom(String assemblyFile)
   at SindaSoft.DependencyWalker.Walker.inspectAssembly(String name)

This was first encountered with version 111.2.70. Also, checked the 'CefSharp.Core.Runtime.dll' version 102.0.100.0. The error is found in this version as well.

-----------------------------------------------------------------
GH question: Are you upgrading from a previous version? No new dependencies have been added in quite some time.

RBO answer: Upgraded from version 102.0.100.0

-----------------------------------------------------------------
GH question: Are you seeing this problem on your development machine or when you deploy to machine?

RBO answer: Sometimes it would show on the development machine, and I don’t know why it’s intermittent. The app with CefSharp would run fine after an initial installation. However, if you copied an updated .exe, after a minor change, to the target machine, the error would show up.

-----------------------------------------------------------------
GH question: Please answer the questions at https://github.com/cefsharp/CefSharp/wiki/Questions-&-Support#when-asking-a-question-please-include-the-following
RBO answer: Answers to questions at link.
1.            Version 112.2.70
2.            WinForms
3.            Not .net Core: Targeting .net framework 4.7.2

Thank You,

Richard B. Oberholz
Engineer
Intent Digital
602-404-0100 Ext: 25
Email: ***@***.******@***.***>
Web:  www.intentdigital.com<http://www.intentdigital.com/>
             www.votelynx.com<http://www.votelynx.com/>

**Reply by amaitland** [reply]
> Get the DependancyWalker.net app from isindicic/DependencyWalker.Net: Show dependencies tree for .NET assemblies like old Depend Walker show it for non managed applications. (github.com)<https://github.com/isindicic/DependencyWalker.Net> and run it against the CefSharp.Core.Runtime.dll.

- Built `DependancyWalker.Net` from source 
- Built `MinimalExample` from source (net472 solution)
- Opened `CefSharp.Core.Runtime.dll` from bin folder of `MinimalExample`, dll loaded without exception, listed dependencies as you'd expect.

> I could not duplicate the problem using the ‘MinimalExample’.

How does your project/solution differ from the `MinimalExample`?

> Checking this .dll with DependencyWalker.net, it reported that the CefSharp.Core.Runtime.dll has the problem.

Try running it against the bin output from the `MinimalExample`.

> An attempt was made to load a program with an incorrect format. usually means you are loading the incorrect architecture dll.

If your application is `x64` then I'd expect this error if it attempted to load an `x86` dll.

You can add some debugging to your build output which might shed some light on what's going on.

### Troubleshooting
Add the following to your `csproj/vbproj` file at the end

```xml
<Target Name="CefSharpAfterBuildDebug" AfterTargets="AfterBuild">
    <CallTarget Targets="CefSharpAfterBuildDiagnostic" />
</Target>
```

Change your `MSBuild project output verbosity` to `Normal`:
```
1. Go to Tools->Options menu in VS
2. Open Projects and Solutions->Build and Run
3. Change the value of the MSBuild project build output verbosity. Pick Normal or higher
```

After you build, check the output window for a block of text that looks like:
```
>Target CefSharpAfterBuildDebug:
>  Target CefSharpAfterBuildDiagnostic:
>    CefSharp After Build Diagnostic
>    CefSharpBuildAction = Content
>    CefSharpPropertiesLoaded = true
>    CefSharpTargetDir =
>    CefSharpTargetDirAnyCpu32 = x86\
>    CefSharpTargetDirAnyCpu64 = x64\
>    RuntimeIdentifier = win7-x64
>    EffectivePlatform =
>    Platform = x64
>    PlatformName = x64
>    Platforms = x86;x64;AnyCPU
>    PlatformTarget = x64
>    CefSharpPlatformTarget = x64
>    PlatformTargetAsMSBuildArchitecture = x64
>    TargetFramework = net472
>    TargetFrameworkVersion = v4.7.2
>    libcef.dll exists = bin.net472\x64\Debug\net472\libcef.dll
>    CefSharpTransformXmlDllPath = C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VisualStudio\v16.0\Web\Microsoft.Web.Publishing.Tasks.dll
>    NuGetProjectStyle = PackageReference
>    NuGetToolVersion = 5.11.1
>    OutputType = WinExe
>    OutDir = bin.net472\x64\Debug\net472\
>    Prefer32Bit = false
```

Copy and paste the output here, making sure to format it for ease of reading.

- https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-text
- https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code

**Comment by rboberholz** [comment]
Alex,

GH Question: How does your project/solution differ from the MinimalExample?
RBO Answer:
I wasn’t totally clear about the results I got when running the MinimalExample. I only ran it and didn’t see any problems. I did use the DependancyWalker.net utility and checked the dependencies of the CefSharp.core.Runtime.dll. I see the same reported dependance issue.
See the following results:

F:\Visual Studio Community Projects\CefSharp.MinimalExample-master\CefSharp.MinimalExample.WinForms\bin.net472\Debug\net472\x64\CefSharp.Core.Runtime.dll - Error loading
System.BadImageFormatException: Could not load file or assembly 'file:///F:\Visual Studio Community Projects\CefSharp.MinimalExample-master\CefSharp.MinimalExample.WinForms\bin.net472\Debug\net472\x64\CefSharp.Core.Runtime.dll' or one of its dependencies. An attempt was made to load a program with an incorrect format.

File name: 'file:///F:\Visual Studio Community Projects\CefSharp.MinimalExample-master\CefSharp.MinimalExample.WinForms\bin.net472\Debug\net472\x64\CefSharp.Core.Runtime.dll'
   at System.Reflection.RuntimeAssembly._nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, Evidence assemblySecurity, RuntimeAssembly reqAssembly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadFrom(String assemblyFile, Evidence securityEvidence, Byte[] hashValue, AssemblyHashAlgorithm hashAlgorithm, Boolean forIntrospection, Boolean suppressSecurityChecks, StackCrawlMark& stackMark)
   at System.Reflection.Assembly.LoadFrom(String assemblyFile)
   at SindaSoft.DependencyWalker.Walker.inspectAssembly(String name)

Alex, please get a copy of DependanceWalker.net and check the Cefsharp.Core.Runtime.dll yourself.


GH Question: If your application is x64 then I'd expect this error if it attempted to load an x86 dll.
                RBO Answer: App architecture is x64.


GH Question: You can add some debugging to your build output which might shed some light on what's going on.
                RBO Answer: This info is here because the the links took me to a page that talked about ‘Basic formatting syntext’ but not a place to upload the text.
The target build info is as follows:

1>CefSharpAfterBuildDiagnostic:
1>  CefSharp After Build Diagnostic
1>  CefSharpBuildAction = Content
1>  CefSharpPropertiesLoaded = true
1>  CefSharpTargetDir =
1>  CefSharpTargetDirAnyCpu32 = x86\
1>  CefSharpTargetDirAnyCpu64 = x64\
1>  CefSharpAnyCpuSupport =
1>  RuntimeIdentifier =
1>  EffectivePlatform =
1>  Platform = x64
1>  PlatformName = x64
1>  Platforms =
1>  PlatformTarget = x64
1>  CefSharpPlatformTarget = x64
1>  PlatformTargetAsMSBuildArchitecture = x64
1>  TargetFramework =
1>  TargetFrameworkVersion = v4.7.2
1>  libcef.dll exists = bin\Debug\libcef.dll
1>  CefSharpTransformXmlDllPath = D:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VisualStudio\v16.0\Web\Microsoft.Web.Publishing.Tasks.dll
1>  NuGetProjectStyle =
1>  NuGetToolVersion =
1>  OutputType = WinExe
1>  OutDir = bin\Debug\
1>  Prefer32Bit = false

All the liks did was take me to a page that talked about ‘Gasic formatting syntext’ but no place to upload the text.

Thank You,

Richard B. Oberholz
Engineer
Intent Digital
602-404-0100 Ext: 25
Email: ***@***.******@***.***>
Web:  www.intentdigital.com<http://www.intentdigital.com/>
             www.votelynx.com<http://www.votelynx.com/>

**Reply by amaitland** [reply]
> Alex, please get a copy of DependanceWalker.net and check the Cefsharp.Core.Runtime.dll yourself.

As per my previous comment, I built from source and tested just fine. The problem you are seeing with `DependencyWalker` is https://github.com/isindicic/DependencyWalker.Net/issues/5

A `32bit` exe attempting to load a `64bit` dll will result in a `BadImageFormatException`. The exception has nothing to do with `CefSharp`.

> GH Question: If your application is x64 then I'd expect this error if it attempted to load an x86 dll. RBO Answer: App architecture is x64.

At face value the exception suggests your `64bit` application is attempting to load a `32bit` dll. How this is happening is not clear.

> The target build info is as follows:

All looks pretty standard. Looks like you are using `packages.config`, you can try migrating to [PacakgeReference](https://learn.microsoft.com/en-us/nuget/consume-packages/migrate-packages-config-to-package-reference) to see if that makes a difference.
